wicker/core/datasets.py

```diff
@@ -125,79 +135,133 @@ def __len__(self) -> int:
-        :param data_service: S3DataService instance to use, defaults to None which uses default initializations
         :param filelock_timeout_seconds: number of seconds after which to timeout on waiting for downloads,
             defaults to FILE_LOCK_TIMEOUT_SECONDS
-        :param treat_objects_as_bytes: If set, don't try to decode ObjectFields and keep them as binary data
+        :param local_cache_path_prefix: Path to local cache path, if empty don't create cache
+        :param treat_objects_as_bytes: If set, don't try to decode ObjectFields and keep them as binary data.
         :param filters: Only returns rows which match the filter. Defaults to None, i.e., returns all rows.
         :type filters: pyarrow.compute.Expression, List[Tuple], or List[List[Tuple]], optional
         .. seealso:: `filters in <https://arrow.apache.org/docs/python/generated/pyarrow.parquet.read_table.html>`__ # noqa
         """
-        super().__init__()
-        self._columns_to_load: Optional[List[str]] = columns_to_load
+        self._arrow_table: Optional[pyarrow.Table] = None  # Set by lazy initialization
+        self._columns_to_load = columns_to_load
+        self._dataset_id = DatasetID(name=dataset_name, version=dataset_version)
+        self._dataset_name = dataset_name
+        self._dataset_version = dataset_version
+        self._filelock_timeout_seconds = filelock_timeout_seconds
+        self._filters = filters
+        self._local_cache_path_prefix = local_cache_path_prefix
+        self._pa_filesystem = pa_filesystem
+        self._partition = DatasetPartition(dataset_id=self._dataset_id, partition=dataset_partition_name)
+        self._path_factory = path_factory
+        self._schema: Optional[DatasetSchema] = None  # Set by lazy initialization
+        self._storage = storage
         self._treat_objects_as_bytes = treat_objects_as_bytes
-        self._schema: Optional[DatasetSchema] = None
-        self._arrow_table: Optional[pyarrow.Table] = None
 
-        self._local_cache_path_prefix = local_cache_path_prefix
-        self._filelock_timeout_seconds = filelock_timeout_seconds
-        self._storage = storage if storage is not None else S3DataStorage()
-        self._s3_path_factory = s3_path_factory if s3_path_factory is not None else S3PathFactory()
-        self._column_bytes_file_cache = ColumnBytesFileCache(
-            local_cache_path_prefix=local_cache_path_prefix,
-            filelock_timeout_seconds=filelock_timeout_seconds,
-            path_factory=self._s3_path_factory,
-            storage=self._storage,
-            dataset_name=dataset_name,
-        )
-        self._pa_filesystem = (
-            pafs.S3FileSystem(region=get_config().aws_s3_config.region) if pa_filesystem is None else pa_filesystem
+        # Create the column bytes reader that depends on the other init parameter values.
+        self._column_bytes_file_reader = self._build_column_bytes_reader()
+
```

Comment:
In the rest of the `BaseDataset` class, it is pretty nice that `__getitem__`, `__len__`, etc. don't need storage-specific logic themselves and can mostly rely on `_storage` instance and `_column_bytes_file_reader`  to provide this.